python/src/uagents/mailbox.py

```diff
@@ -1,101 +1,171 @@
-from websockets import connect
+from pydantic import UUID4, BaseModel, ValidationError
 
-from uagents.config import MAILBOX_POLL_INTERVAL_SECONDS
-from uagents.crypto import is_user_address
+from uagents.config import MAILBOX_POLL_INTERVAL_SECONDS, AgentType, AgentverseConfig
+from uagents.crypto import Identity, is_user_address
 from uagents.dispatch import dispatcher
 from uagents.envelope import Envelope
+from uagents.types import AgentEndpoint
 from uagents.utils import get_logger
 
+logger = get_logger("mailbox")
 
-class MailboxClient:
-    """Client for interacting with the Agentverse mailbox server."""
 
-    def __init__(self, agent, logger: Optional[logging.Logger] = None):
-        self._agent = agent
-        self._access_token: Optional[str] = None
-        self._envelopes_to_delete = asyncio.Queue()
-        self._poll_interval = MAILBOX_POLL_INTERVAL_SECONDS
-        self._logger = logger or get_logger("mailbox")
+class AgentverseUserToken(BaseModel):
+    value: str
 
-    @property
-    def base_url(self):
-        """
-        Property to access the base url of the mailbox server.
 
-        Returns: The base url of the mailbox server.
+class ChallengeRequest(BaseModel):
+    address: str
 
-        """
-        return self._agent.mailbox["base_url"]
 
-    @property
-    def agent_mailbox_key(self):
-        """
-        Property to access the agent_mailbox_key of the mailbox server.
+class ChallengeResponse(BaseModel):
+    challenge: str
 
-        Returns: The agent_mailbox_key of the mailbox server.
-        """
-        return self._agent.mailbox["agent_mailbox_key"]
 
-    @property
-    def protocol(self):
-        """
-        Property to access the protocol of the mailbox server.
+class RegistrationRequest(BaseModel):
+    address: str
+    challenge: str
+    challenge_response: str
+    agent_type: Optional[AgentType] = None
+    endpoints: Optional[list[AgentEndpoint]] = None
+    agent_mailbox_key: Optional[str] = None
 
-        Returns: The protocol of the mailbox server {ws, wss, http, https}.
-        """
-        return self._agent.mailbox["protocol"]
 
-    @property
-    def http_prefix(self):
-        """
-        Property to access the http prefix of the mailbox server.
+class RegistrationResponse(BaseModel):
+    access_token: Optional[str] = None
+    expiry: Optional[datetime] = None
+
+
+class StoredEnvelope(BaseModel):
+    uuid: UUID4
+    envelope: Envelope
+    received_at: datetime
+    expires_at: datetime
 
-        Returns: The http prefix of the mailbox server {http, https}.
-        """
-        return self._agent.mailbox["http_prefix"]
+
+async def register_in_agentverse(
+    user_token: AgentverseUserToken,
+    identity: Identity,
+    endpoints: list[AgentEndpoint],
+    agentverse: AgentverseConfig,
+) -> Optional[RegistrationResponse]:
+    """
+    Registers agent in Agentverse
+    """
+    async with aiohttp.ClientSession() as session:
+        # get challenge
```

Comment:
Why do we need this challenge? We want to prove that the caller of the registration endpoint is in fact the agent owning the private key to the given address? 
Wouldn't it be sufficient to simply sign the RegistrationRequest below instead of the challenge to save one round trip? We would still be able to verify if the signature matches the given address. A replay attack would only make sense for an adversary if either address or endpoint could be altered, which they can't because they would be signed.

Reply:
It's possible that another approach could also work here, but I'd suggest we keep this one for now and maybe revisit later on whether a simpler method can provide sufficient security.